## Final Project Submission

Please fill out:
* Student name: **Gichogu Macharia**
* Student pace:  **full time**
* Scheduled project review date/time: **23rd OCT 2023**
* Instructor name: **Nikita Njoroge**


# King County House Project

## Business Understanding

In the indutry of real estate the professionals require an approach that avoids relying on intuition but rather hard data to improve on sales and market value. Equiped with the right analytic-driven insights, agencies can improve customer service and gain a competitive advantage.

## Data Understanding

The data provided is from King County House Sales. We will view the information to gain better insight.

In [1]:
#importing the pandas package as pd to read the data
import pandas as pd 

#reading the data as a CSV file
df = pd.read_csv('data\kc_house_data.csv')

#displayng asll columns
pd.set_option('display.max_columns', None)

In [2]:
#verifying the rows and columns
df.shape

(21597, 21)

From the data we can see that we have 21,597 rows and 21 columns.

In [3]:
#looking into the first 5 rows of data
df.head(3)

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   

   sqft_lot  floors waterfront  view condition          grade  sqft_above  \
0      5650     1.0        NaN  NONE   Average      7 Average        1180   
1      7242     2.0         NO  NONE   Average      7 Average        2170   
2     10000     1.0         NO  NONE   Average  6 Low Average         770   

  sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0           0.0      1955           0.0    98178  47.5112 -122.257   
1         400.0      1951        1991.0    98125  47.7210 -122.319   
2           0.0      1933           NaN    98028  47.7379 -122.233   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062

The information provided above reveals our 20 columns which are; 
1. the id number 
2. date
3. price of the house
4. number of bedrooms 
5. number of bathrooms 
6. the square feet of the house 
7. floors
8. evidence of a waterfront 
9. the view, condition of the house 
10. the grade of the house
11. the year the house was built 
12. the year the house was renovated 
13. the zipcode
14. latitutes of the house 
15. longitudes of the house
The first column is the indexing column which I did not list my observations hence removing one column in my assesment.

In [4]:
df.grade.value_counts().sort_values(ascending=False)

7 Average        8974
8 Good           6065
9 Better         2615
6 Low Average    2038
10 Very Good     1134
11 Excellent      399
5 Fair            242
12 Luxury          89
4 Low              27
13 Mansion         13
3 Poor              1
Name: grade, dtype: int64

In [5]:
#check the data information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

Our data types are comprised of intergers, floats and objects. 

We can now move to the next step which is the data analysis.

## Data Preparation

So far we are familiar with our data but we need to make it ready for utilization in order to work with refined data. Or the data we find necessary. First we'll take a look at the first 3 rows of data.

In [6]:
#first 3 rows of data
df.head(3)

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   

   sqft_lot  floors waterfront  view condition          grade  sqft_above  \
0      5650     1.0        NaN  NONE   Average      7 Average        1180   
1      7242     2.0         NO  NONE   Average      7 Average        2170   
2     10000     1.0         NO  NONE   Average  6 Low Average         770   

  sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0           0.0      1955           0.0    98178  47.5112 -122.257   
1         400.0      1951        1991.0    98125  47.7210 -122.319   
2           0.0      1933           NaN    98028  47.7379 -122.233   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062

Based on our objectives the best fit to initiate progress is to remove unnecessary data that we will not use. The data that is to be excluded is;
1. id - We do not need the id to relate the information.
2. date - The date is not a necessity in the analysis.
3. square feets of rooms - Our plan with the data does not include a need for the square feet of rooms.
4. zipcode - Longitues and latitudes are effective enough.
5. the year built and renovated - Our model can work without the years.

In [7]:
#creating a list of columns to drop
#columns_to_drop = ['id', 'date', 'zipcode', 'yr_built']

#removing the columns
df = df.drop(columns=columns_to_drop)

NameError: name 'columns_to_drop' is not defined

In [9]:
occurrence_counts = df.groupby(['zipcode', 'waterfront']).size().reset_index(name='count')
similar_combinations = occurrence_counts[occurrence_counts['count'] > 1]

# Display the similar combinations and their counts
print(similar_combinations)

    zipcode waterfront  count
0     98001         NO    328
1     98002         NO    179
2     98003         NO    255
3     98004         NO    282
5     98005         NO    158
..      ...        ...    ...
93    98178        YES      9
94    98188         NO    119
95    98198         NO    233
96    98198        YES      9
97    98199         NO    275

[92 rows x 3 columns]


In [11]:
# Group the data by 'zipcode' and count occurrences of different 'waterfront' values
occurrence_counts = df.groupby(['zipcode', 'waterfront']).size().reset_index(name='count')

# Filter for combinations with the same 'zipcode' and different 'waterfront'
different_waterfront_combinations = occurrence_counts[occurrence_counts['count'] > 1]

# Display the different combinations and their counts
print(different_waterfront_combinations)
total_count = different_waterfront_combinations['count'].sum()
print(total_count)

    zipcode waterfront  count
0     98001         NO    328
1     98002         NO    179
2     98003         NO    255
3     98004         NO    282
5     98005         NO    158
..      ...        ...    ...
93    98178        YES      9
94    98188         NO    119
95    98198         NO    233
96    98198        YES      9
97    98199         NO    275

[92 rows x 3 columns]
19214


In [ ]:
#confirming that the columns are dropped
df.head(3)

### Cleaning of Data

Working with data requires acuracy and thus means that to get accurate data the data must be cleaned through eliminating errors or extreme data and correcting the incorrect data. 

In [ ]:
#checking the data information
df.info()

#### Creating a refined grade table

The grade is inclusive of the number and at the same time a condition. To give easier data it is better to divide the grade from the condition in the grade column

In [ ]:
#split the grade column to the numeric grade and the condition
df[['grade_no', 'grade_condition']] = df['grade'].str.split(' ', n=1, expand=True)

In [ ]:
#confirming the columns are added
df.head()

In [ ]:
#convert the numeric grade to an interger
df['grade_no'] = df['grade_no'].astype(int)

In [ ]:
#confirming that the grades are divided into intergers and objects 
df.info()

In [ ]:
#confirm that there are no missing values
df['grade_no'].isna().sum()

In [ ]:
#confirm that there are no missing values
df['grade_condition'].isna().sum()

In [ ]:
#drop the original grade column
df = df.drop(columns=['grade'])

In [ ]:
#reviewing the new table
df.head(3)

#### Outliers or Extreme data

The prices might have outliers and give an innacurate reading to our mode. To solve this issue we will remove the extreme values(outliers) for better balance of data

First we need to check for outliers

In [ ]:
#viewing the data information
df.info()

From the data the only place we need to look out for outliers are the prices but the rest can still be functional as they are. In this step we will only concentrate on the prices. We start by defining a function that we can use to check for outliers and call out the function in order to confirm that we indeed have outliers.

In [ ]:
#import matplotlib package to draw a graph
import matplotlib.pyplot as plt

#define a function for removing outliers
def check_outliers(data, column):
    #plotting the graph
    plt.boxplot(data[column])
    
    #setting the title of the plot
    plt.title(f'Box plot of {column}')
    
    #visualizing the plot
    plt.show()
    return

In [ ]:
#checking for the price outlier using the function
check_outliers(df, 'price')

From our observation above we can see that we have outliers in our data. The next step is to remove the outliers so that we can work with a balanced dataset. In this step we will also define a function that can assist us in the removal of our outliers.

In [ ]:
#defining a function to remove outliers
def remove_outliers(data, column):
    #calculate the first quartile(Q1)
    Q1 = data[column].quantile(0.25)
    
    #calculate the third quartile(Q3)
    Q3 = data[column].quantile(0.75)
    
    #calculate the Interquartile Range (IQR)
    IQR = Q3 - Q1
    
    #Define the lower and upper bounds
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    #Remove outliers
    data = data[(data[column] >= lower_bound) & (data[column] <= upper_bound)]
    return data

In [ ]:
#remove the price outlier
df = remove_outliers(df, 'price')

Let's confirm if the outliers have been removed from the data.

In [ ]:
#confirming that the outliers are removed
check_outliers(df, 'price')

From the figure above we can indeed conclude that the outliers are removed and now we can work with a balanced dataset.

The other thing that we can conclude is the data is cleaned and is ready for analysis. This means that our data can now be used for the next step which is the modelling.

In [ ]:
#create a function that removes null values
#def remove_null(data, column):
    #data.dropna(subset=[column], inplace=True)
    #return data

## Modelling 